packages/@react-spectrum/tag/src/TagGroup.tsx

```diff
@@ -10,50 +10,88 @@
  * governing permissions and limitations under the License.
  */
 
-import {classNames, useStyleProps} from '@react-spectrum/utils';
-import React, {useContext} from 'react';
-import {Removable} from '@react-types/shared';
+import {classNames, useDOMRef, useStyleProps} from '@react-spectrum/utils';
+import {DOMRef} from '@react-types/shared';
+import {GridCollection, useGridState} from '@react-stately/grid';
+import {GridKeyboardDelegate, useGrid} from '@react-aria/grid';
+import {mergeProps} from '@react-aria/utils';
+import React, {ReactElement, useMemo} from 'react';
 import {SpectrumTagGroupProps} from '@react-types/tag';
 import styles from '@adobe/spectrum-css-temp/components/tags/vars.css';
+import {Tag} from './Tag';
+import {useListState} from '@react-stately/list';
+import {useLocale} from '@react-aria/i18n';
 import {useProviderProps} from '@react-spectrum/provider';
 import {useTagGroup} from '@react-aria/tag';
 
-interface TagGroupContextValue extends Removable<any, void> {
-  isDisabled?: boolean,
-  isFocused?: boolean,
-  isRequired?: boolean,
-  isReadOnly?: boolean,
-  validationState?: 'valid' | 'invalid',
-  role?: 'gridcell'
-}
-
-const TagGroupContext = React.createContext<TagGroupContextValue | {}>({});
-
-export function useTagGroupProvider(): TagGroupContextValue {
-  return useContext(TagGroupContext);
-}
 
-export const TagGroup = ((props: SpectrumTagGroupProps) => {
+function TagGroup<T extends object>(props: SpectrumTagGroupProps<T>, ref: DOMRef<HTMLDivElement>) {
   props = useProviderProps(props);
-
   let {
-    isReadOnly,
     isDisabled,
+    isRemovable,
     onRemove,
-    validationState,
-    children,
     ...otherProps
   } = props;
+
+  let domRef = useDOMRef(ref);
   let {styleProps} = useStyleProps(otherProps);
+  let {direction} = useLocale();
   const {tagGroupProps} = useTagGroup(props);
+  let {collection} = useListState(props);
+  let gridCollection = useMemo(() => new GridCollection({
+    columnCount: isRemovable ? 2 : 1,
+    items: [...collection].map(item => {
+      let childNodes = [{
+        ...item,
+        index: 0,
+        type: 'cell'
+      }];
 
-  function removeAll(tags) {
-    onRemove([tags]);
-  }
+      // add column of clear buttons if removable
+      if (isRemovable) {
+        childNodes.push({
+          key: `remove-${item.key}`,
+          type: 'cell',
+          index: 1,
+          value: null,
+          level: 0,
+          rendered: null,
+          textValue: item.textValue, // TODO localize?
+          hasChildNodes: false,
+          childNodes: []
+        });
+      }
+
+      return {
+        type: 'item',
+        childNodes
+      };
+    })
+  }), [collection, isRemovable]);
+
+  let state = useGridState({
+    ...props,
+    collection: gridCollection,
+    focusMode: 'cell'
```

Comment:
```suggestion
    focusMode: 'row'
```
This should fix the issue that I noted previously with the treewalker blowing up. I did notice that the Tags aren't focused when clicked, looking into that right now

Reply:
Additionally, I noticed that the TagGroup div flickers with a blue focus ring before focus settles on one of the Tags. May need to add something in the css to ensure it doesn't get a focus ring

Reply:
As for the Tags not being focused when clicked, you'll probably need to add a `usePress` in `useTag` and merge the props in `tagProps`